starlite/types/parsed_signature.py

```diff
@@ -0,0 +1,187 @@
+        return self.origin and issubclass(self.origin, Collection)
+
+    @classmethod
+    def from_annotation(cls, annotation: Any) -> ParsedType:
+        """Initialize ParsedType.
+
+        Args:
+            annotation: The type annotation. This should be extracted from the return of
+                ``get_type_hints(..., include_extras=True)`` so that forward references are resolved and recursive
+                ``Annotated`` types are flattened.
+
+        Returns:
+            ParsedType
+        """
+        unwrapped, metadata, wrappers = unwrap_annotation(annotation)
+
+        origin = get_origin(unwrapped)
+        args = get_args(unwrapped)
+        return ParsedType(
+            raw=annotation,
+            annotation=unwrapped,
+            origin=origin,
+            args=args,
+            metadata=metadata,
+            is_annotated=Annotated in wrappers,
+            is_required=Required in wrappers,
+            is_not_required=NotRequired in wrappers,
+            safe_generic_origin=get_safe_generic_origin(origin),
+            inner_annotations=tuple(cls.from_annotation(arg) for arg in args),
+        )
+
+
+@dataclass(frozen=True)
+class ParsedParameter:
+    """Represents the parameters of a callable."""
+
+    name: str
+    """The name of the parameter."""
+    default: Any | Empty
+    """The default value of the parameter."""
+    parsed_type: ParsedType
+    """The annotation of the parameter."""
+
+    @property
+    def annotation(self) -> Any:
+        """The annotation of the parameter."""
+        return self.parsed_type.annotation
+
+    @property
+    def has_default(self) -> bool:
+        """Whether the parameter has a default value or not."""
+        return self.default is not Empty
+
+    @classmethod
+    def from_parameter(cls, parameter: Parameter, fn_type_hints: dict[str, Any]) -> ParsedParameter:
+        """Initialize ParsedSignatureParameter.
+
+        Args:
+            parameter: inspect.Parameter
+            fn_type_hints: mapping of names to types. Should be result of ``get_type_hints()``, preferably via the
+            :attr:``get_fn_type_hints() <.utils.signature_parsing.get_fn_type_hints>` helper.
+
+        Returns:
+            ParsedSignatureParameter.
+        """
+        try:
+            annotation = fn_type_hints[parameter.name]
+        except KeyError as err:
+            raise ImproperlyConfiguredException(
+                f"'{parameter.name}' does not have a type annotation. If it should receive any value, use 'Any'."
+            ) from err
+
+        if parameter.name == "state" and not issubclass(annotation, ImmutableState):
+            raise ImproperlyConfiguredException(
+                f"The type annotation `{annotation}` is an invalid type for the 'state' reserved kwarg. "
+                "It must be typed to a subclass of `starlite.datastructures.ImmutableState` or "
+                "`starlite.datastructures.State`."
+            )
```

Comment:
I think this should go into the `BaseRouteHandler._validate_handler_function`. It logically belongs there, since this is where we do the other signature validations as well. It also makes `ParsedParameter` more generic.

Reply:
I agree that this doesn't belong here - keeping this stuff as generic as possible is ideal. However, putting it in the handler validation only would mean that dependency functions aren't similarly validated as they are now.

I'll create a utility and look into adding a `Provide.on_registration()` callback as well, or something along those lines.

Reply:
I'm going to make a separate issue for resolving this.